### PR TITLE
Support for slideshow mode

### PR DESCRIPTION
**Pull request by 3lvis**
Enables starting the Viewer in *slideshow* mode, basically automatically goes to the next image after a certain period of time.

## Usage

```swift
let viewerController = ViewerController(initialIndexPath: indexPath, collectionView: collectionView, isSlideshow: true)
viewerController.dataSource = self
viewerController.delegate = self
present(viewerController, animated: false, completion: nil)
```

slideshow

### CODE REVIEW

File: Library/Photo.swift [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import UIKit
 import Photos
 
-struct Section {
+class Section {
```
Comment (marijnschilling) [inline Library/Photo.swift:4]: Why did you change this into a class? (just curious)

File: Library/Photo.swift [hunk 2]
```diff
@@ -10,7 +10,7 @@ struct Section {
     }
 }
 
-struct Photo: Viewable {
+class Photo: Viewable {
```
Comment (marijnschilling) [inline Library/Photo.swift:13]: Why did you change this into a class? (just curious)

File: Library/PhotosController.swift [hunk 3]
```diff
@@ -48,9 +47,31 @@ class PhotosController: UICollectionViewController {
             self.sections = Photo.constructRemoteElements()
             self.collectionView?.reloadData()
         }
+
+        #if os(tvOS)
+        let playPauseTapRecognizer = UITapGestureRecognizer(target: self, action: #selector(self.playPause(gesture:)))
+        playPauseTapRecognizer.allowedPressTypes = [NSNumber(value: UIPressType.playPause.rawValue)]
```
Comment (marijnschilling) [inline Library/PhotosController.swift:53]: What is this `allowedPressTypes`? What kind of gesture on the collectionview will activate the slideshow? Might be more clear to put the pressType var in a value with a clearer name, before setting it as allowedPressTypes`

File: Library/PhotosController.swift [hunk 4]
```diff
@@ -48,9 +47,31 @@ class PhotosController: UICollectionViewController {
...
+        self.collectionView?.addGestureRecognizer(playPauseTapRecognizer)
+        #endif
     }
 
     #if os(tvOS)
+    func playPause(gesture: UITapGestureRecognizer) {
```
Comment (marijnschilling) [inline Library/PhotosController.swift:59]: Oke s here you start playing the slideShow right? From the focussed cell onwards. Im just wondering how it will pause in this method...

File: Library/PhotosController.swift [hunk 5]
```diff
@@ -70,6 +91,13 @@ class PhotosController: UICollectionViewController {
 
         return alertController
     }
+
+    func object(at indexPath: IndexPath) -> Photo {
```
Comment (marijnschilling) [inline Library/PhotosController.swift]: If this method always return a `Photo` why isnt it called `photo(at indexPath:)`?

File: Source/SlideshowView.swift [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+class SlideshowView: UIView, ViewableControllerContainer {
+    weak var dataSource: ViewableControllerContainerDataSource?
+    weak var delegate: ViewableControllerContainerDelegate?
+    fileprivate unowned var parentController: UIViewController
+    fileprivate var currentPage: Int
+    fileprivate var currentController: ViewableController?
+
+    fileprivate lazy var timer: Timer = {
+        let timer = Timer(timeInterval: 6, target: self, selector: #selector(loadNext), userInfo: nil, repeats: true)
```
Comment (marijnschilling) [inline Source/SlideshowView.swift]: I would make this into a static let of this class

File: Source/SlideshowView.swift [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        self.loadPage(self.currentPage, animated: false, isInitial: true)
+    }
+
+    fileprivate func loadPage(_ page: Int, animated: Bool, isInitial: Bool) {
+        let numPages = self.dataSource?.numberOfPagesInViewableControllerContainer(self) ?? 0
+        if page >= numPages || page < 0 {
+            return
+        }
+
+        guard let controller = self.dataSource?.viewableControllerContainer(self, controllerAtIndex: page) as? ViewableController, controller.view.superview == nil else { return }
```
Comment (marijnschilling) [inline Source/SlideshowView.swift]: So it get's the controller at the indexPath of the collectionview form the superview? Or am I mistaken?

File: Source/SlideshowView.swift [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        let hasReachedEnd = newPage == numPages
+        if hasReachedEnd {
+            newPage = 0
+        }
+
+        self.loadPage(newPage, animated: true, isInitial: false)
+    }
+
+    func start() {
+        RunLoop.current.add(self.timer, forMode: .defaultRunLoopMode)
```
Comment (marijnschilling) [inline Source/SlideshowView.swift:92]: What is this RunLoop thing?

**Review by marijnschilling — COMMENTED** [review]
I mostly asked questions, but I think it looks good! Also god to know what you where working on from the iOS meeting yesterday ✌️

File: Library/Photo.swift (hunk 1 above)
Comment (3lvis) [reply Library/Photo.swift:4]: [Here](https://github.com/bakkenbaeck/Viewer/blob/41f1eba26b8ed7cee6d21deb09e98e6bfda6c02c/Library/PhotosController.swift#L159-L166) I'm modifying an element from the array, if it's a struct the original element doesn't get modified, just a copy of it, so you have to insert it back into the array and remove the old copy. This was easily avoided using `class`, then any modification to an object results in the actual object being modified.

screen shot 2017-04-19 at 11 37 25

File: Library/Photo.swift (hunk 2 above)
Comment (3lvis) [reply Library/Photo.swift:13]: Same reason as above ☝️

File: Library/PhotosController.swift (hunk 3 above)
Comment (3lvis) [reply Library/PhotosController.swift:53]: It's an Apple TV thingie. When you press "playPause" it starts the slideshow. You'll find this pattern everywhere when searching for how to interact with the Siri Remote.

https://developer.apple.com/library/content/documentation/General/Conceptual/AppleTV_PG/DetectingButtonPressesandGestures.html

File: Library/PhotosController.swift (hunk 4 above)
Comment (3lvis) [reply Library/PhotosController.swift:59]: There's no pause for a slideshow, you press "play" for starting the slideshow and press "Menu" (it means Home/Back in the Apple TV) to stop the slideshow.

File: Library/PhotosController.swift (hunk 5 above)
Comment (3lvis) [reply Library/PhotosController.swift]: Could be called photo as well :)

File: Source/SlideshowView.swift (hunk 6 above)
Comment (3lvis) [reply Source/SlideshowView.swift]: Done!

File: Source/SlideshowView.swift (hunk 8 above)
Comment (3lvis) [reply Source/SlideshowView.swift:92]: This is the way you start timers.

File: Source/SlideshowView.swift (hunk 7 above)
Comment (3lvis) [reply Source/SlideshowView.swift]: It asks for a controller at a certain index, the second check is just to make sure that the controller wasn't already added as a subview.

Comment (3lvis) [reply Source/SlideshowView.swift]: I don't think the second check is needed, gonna remove it.

File: Library/PhotosController.swift (hunk 3 above)
Comment (marijnschilling) [reply Library/PhotosController.swift:53]: Oke so .playPause is a astandard enum thngy that you turn unto an NSNumber and add to an array to set as `allowedPressTypes` 👍

File: Library/PhotosController.swift (hunk 4 above)
Comment (marijnschilling) [reply Library/PhotosController.swift:59]: Cool makes sense now that I know 'playPause' is a standard AppleTV button

**Comment by marijnschilling** [comment]
Oke thanks for explaining I think it looks good! I say 👕 -it (hahaha i mean ship-it) unless you have anything else you want me to look at 👀

**Comment by 3lvis** [comment]
@marijnschilling Thanks for reviewing! I need to do some pre-loading but I'll do it in a future PR.
